java/org/apache/catalina/realm/DataSourceRealm.java

```diff
@@ -539,6 +612,162 @@ private boolean isRoleStoreDefined() {
     }
 
 
+    /**
+     * Return the specified user's requested user attributes as a map.
+     * 
+     * @param dbConnection The database connection to be used
+     * @param username User name for which to return user attributes
+     * 
+     * @return a map containing the specified user's requested user attributes
+     */
+    protected Map<String, Object> getUserAttributesMap(Connection dbConnection, String username) {
+
+        String preparedAttributes = getUserAttributesStatement(dbConnection);
+        if (preparedAttributes == null || preparedAttributes == USER_ATTRIBUTES_NONE_REQUESTED) {
+            // The above reference comparison is intentional. USER_ATTRIBUTES_NONE_REQUESTED
+            // is a tag object (empty String) to distinguish between null (not yet
+            // initialized) and empty (no attributes requested).
+            // TODO Could as well be changed to `preparedAttributes.lenghth() = 0`
+
+            // Return null if no user attributes are requested (or if the statement was not
+            // yet built successfully)
+            return null;
+        }
+
+        try (PreparedStatement stmt = dbConnection.prepareStatement(preparedAttributes)) {
+            stmt.setString(1, username);
+
+            try (ResultSet rs = stmt.executeQuery()) {
+
+                if (rs.next()) {
+                    Map<String, Object> attrs = new LinkedHashMap<>();
+                    ResultSetMetaData md = rs.getMetaData();
+                    int ncols = md.getColumnCount();
+                    for (int columnIndex = 1; columnIndex <= ncols; columnIndex++) {
+                        String columnName = md.getColumnName(columnIndex);
+                        // Ignore case, database may have case-insensitive field names
+                        if (columnName.equalsIgnoreCase(userCredCol)) {
+                            // Always skip userCredCol (must be there if all columns
+                            // have been requested)
+                            continue;
+                        }
+                        attrs.put(columnName, rs.getObject(columnIndex));
+                    }
+                    return attrs.size() > 0 ? attrs : null;
```

Comment:
Some user stores may have multivalued attributes. For this realm you should mention that it only supports single-valued ones.

Reply:
You're thinking of array-typed columns? Actually, these don't work as expected. `getObject(int)` only returns a string representation of the array. Maybe that's what the documentation means by _materialized_

>  In the JDBC 2.0 API, the behavior of method `getObject` is extended to materialize data of SQL user-defined types. 

I'm trying to find a way to get real arrays from the database with standard JDBC (in contrast to using driver proprietary types and methods).

What other types of multi-valued attributes are you thinking of?

Reply:
Ah, applies as well to BLOBs, CLOBs and the like.

Reply:
Not array-typed columns. A query might return more than one line per user attribute which would actually override all previous ones with the last one. I don't expect support for this here, but just like to mention it. If you don't want to support that is perfectly fine, but it should simply be documented that's it.

Reply:
> What other types of multi-valued attributes are you thinking of?

I haven't use SQL databases for any user store for the last 15 years, but only Active Directory so I am used to multi-valued values.

Reply:
I'm about to implement support for (at least) arrays, BLOBS and CLOBS. Likely it is sufficient to support one-dimensional arrays only. Everything _not_ supported will be documented (both in Javadoc comments as well as in Tomcat's configuration reference).

Reply:
Ah, OK, multiple records for a user... Since by definition, the attributes are extra fields of the _user table_ (that has logon name and password), there is generally no support for user tables that contain the user more than once. That's just not allowed. In method `getPassword`  credentials are read from the first record returned (there's no `while (rs.hasNext())` loop).

So generally, having a user more than once in the user table may cause problems with attributes: in `getUserAttributesMap` the multiple rows for that username could come in a different order. Than, the Principal will get the wrong attributes.

We could generally deny authenticating in a user, for which more than one row is present in the user table (as done in JNDIRealm, AFAIK). However, that has nothing to do with the enhancement and should probably got to a new PR.

As an alternative, we could add an ORDER BY <credential_field> to both the _credential_ and the _attributes_ statements in order to force the same order of rows.

Reply:
Yes, I completely forgot the row orientation on the database. No need to add support since this would require a schema change. Don't bother, you already did a lot.

Reply:
I'm almost done with support for SQL-Arrays, BLOBs and CLOBS. I thought that these come with `getObject()` for free. But, a couple of lines of code are required to get these types as well.

Arrays are a quite cool feature and provide multi-valued attribute support. BLOBS may be quite useful to provide binary data, like images and similar stuff.